### PR TITLE
Introduce "plugins" option

### LINKED ISSUE #45: Add option to run a processor on imported AST
Convenient to run a linter on the fly. No extra processing.
Ideas:
- change onImport option to pass all files + latest filename/ast
- add an afterParse (or onParse) option that take filename + ast

### PR DESCRIPTION
**Pull request by jedmao**
Fixes #45

Willing to rename this option to something else, but I would appreciate your code review, @MoOx. I think I still need to write one more test that verifies warnings are being displayed for plugins applied on import.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -71,22 +76,33 @@ function AtImport(options) {
...
+    })
+  }
+}
+
+function createProcessor(result, plugins) {
+  if (plugins) {
+    if (Array.isArray(plugins)) {
+      return postcss(plugins)
     }
+    result.warn("plugins must be an array")
```
Comment (MoOx) [inline index.js]: We should probably throw an error instead for this.

File: test/index.js [hunk 2]
```diff
@@ -54,10 +59,16 @@ test("@import", function(t) {
...
 
   compareFixtures(t, "transform", "should support transform", {
-    path: importsDir,
     transform: require("css-whitespace"),
   })
 
+  compareFixtures(t, "plugins", "should support plugins", {
+    plugins: [
+      require("postcss-nested"),
+      require("postcss-bem-linter"),
```
Comment (MoOx) [inline test/index.js]: instead of adding dev deps (real plugins) we should code & use a stupid plugin with more stupid fixtures

**Comment by MoOx** [comment]
I merged another pr so you might need to be rebased on master.
This looks good.
I am not sure why you did change lots of others tests (by using postcss async api, which make code harder to read).
Also tell me if I am wrong, but with your modification, we are not far from #41 ?

I am just afraid that with the change in compareFixtures function, you didn't update all t.end() called after the call to that function which might miss some unfinished tests.

File: index.js (hunk 1 above)
Comment (jedmao) [inline index.js]: agreed

File: test/index.js (hunk 2 above)
Comment (jedmao) [inline test/index.js]: will do

**Comment by jedmao** [comment]
@MoOx maybe I can decompose this work into 3 separate PRs:
- The plugins option
- The postcss async api
- Asynchronous file I/O #41

Stay tuned.

**Comment by MoOx** [comment]
cool. Can't wait for those things :)

**Comment by jedmao** [comment]
@MoOx I've updated this PR with the following things:

1. Rebased on master.
1. Found one `t.end()` that wasn't being called properly and fixed it. I also made use of tape's `t.plan(n)` method to ensure the number of tests we expect to run are actually being run. Currently, master is running 25 tests in total. I've added 3 tests for a new total of 28, so I'm sure nothing is left behind.
1. Removed the real plugins and replaced with stupid plugins.
1. Throw an error when plugins option is not an array.
1. Squashed changes.

I want to explain why I changed this to use the async API. Actually, I tried keeping things sync at first with `postcss().process(newStyles).css`, but I got errors because some of the plugins in my plugins option were asynchronous; namely, [postcss-bem-linter](https://github.com/postcss/postcss-bem-linter/blob/333f84fd0336b457793ec7d6a817adb87fdd6bb1/index.js#L40). Accepting this fact, I went async. Things got a little weird at that point and the tests were throwing a lot of errors, which I was able to resolve by making this entire plugin async. Of course, this means the tests also had to be modified to be async in order to work.

You're right that we're closer to #41 with this PR.

**Comment by jedmao** [comment]
Also, I tried decomposing this PR into two separate PRs, but I couldn't get it to work without this plugin being async. That's why it's all still in the same PR.

**Comment by jedmao** [comment]
@MoOx you might also find this PR easier to review by [removing whitespace](https://github.com/postcss/postcss-import/pull/55/files?w=1).

**Comment by MoOx** [comment]
You did a great job here ! (I know the whitespace trick, thanks ;)

I just see one last detail https://github.com/postcss/postcss/blob/master/docs/api.md#resultwarntext-opts
The api expect some text, but a warning returned by warnings() is an object. So maybe we should just concat the `result.messages` array.

I will do that and add proper test to ensure messages are ok :)

**Comment by MoOx** [comment]
Released as 6.1.0

**Comment by jedmao** [comment]
Your release notes say it's a `plugin` option in the singular, but it's `plugins` in the plural.

**Comment by MoOx** [comment]
Oh my bad, I will fix that :) Thanks!

**Pull request merged**